### ISSUE TITLE
retention service  was blocking when Close TSMReader

### ISSUE THREAD

**Issue by GenuineJyn** [influxdata/influxdb#16044]
v1.6.1  
```
177271 goroutine 80058 [semacquire, 41921 minutes]:
177272 sync.runtime_Semacquire(0xc045f20710)
177273     /usr/local/Cellar/go/1.11.5/libexec/src/runtime/sema.go:56 +0x39
177274 sync.(*WaitGroup).Wait(0xc045f20708)
177275     /usr/local/Cellar/go/1.11.5/libexec/src/sync/waitgroup.go:130 +0x64
177276 github.com/influxdata/influxdb/tsdb/engine/tsm1.(*TSMReader).Close(0xc045f20700, 0x0, 0x0)
177277     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/reader.go:364 +0x3d
177278 github.com/influxdata/influxdb/tsdb/engine/tsm1.(*FileStore).Close(0xc00c7ec4e0, 0x133e3e8, 0xc00058eb40)
177279     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/file_store.go:604 +0xd0
177280 github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).Close(0xc00058eb40, 0x0, 0x0)
177281     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:768 +0xa0
177282 github.com/influxdata/influxdb/tsdb.(*Shard).close(0xc0074fc6c0, 0x133e3e8, 0xc0074fc878)
177283     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/shard.go:378 +0x47
177284 github.com/influxdata/influxdb/tsdb.(*Shard).Close(0xc0074fc6c0, 0x0, 0x0)
177285     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/shard.go:368 +0x79
177286 github.com/influxdata/influxdb/tsdb.(*Store).DeleteShard(0xc000185000, 0x54a, 0x0, 0x0)
177287     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/store.go:780 +0x317
177288 github.com/influxdata/influxdb/services/retention.(*Service).run(0xc002abf0e0)
177289     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/services/retention/service.go:134 +0x261
177290 github.com/influxdata/influxdb/services/retention.(*Service).Open.func1(0xc002abf0e0)
177291     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/services/retention/service.go:51 +0x57
177292 created by github.com/influxdata/influxdb/services/retention.(*Service).Open
177293     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/services/retention/service.go:51 +0x1ae
```

**Comment by russorat** [comment]
@GenuineJyn thanks for reporting! could you retry with the latest 1.7.9 release? If the issue is still there, could you add some repro steps to help the team investigate?

**Comment by GenuineJyn** [comment]
So far, For some reasons we can only use this version temporarily. one_month rp has been created before. After more than one month，we found that the expired shardgroups were still here.  The above logs were found from `*:8086debug/pprof`. when drop shards manually,  some shards also failed, logs as bellow:

```
218941 goroutine 21798232784 [semacquire, 3792 minutes]:
218942 sync.runtime_Semacquire(0xc03ec10400)
218943     /usr/local/Cellar/go/1.11.5/libexec/src/runtime/sema.go:56 +0x39
218944 sync.(*WaitGroup).Wait(0xc03ec103f8)
218945     /usr/local/Cellar/go/1.11.5/libexec/src/sync/waitgroup.go:130 +0x64
218946 github.com/influxdata/influxdb/tsdb/engine/tsm1.(*TSMReader).Close(0xc03ec103f0, 0x0, 0x0)
218947     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/reader.go:364 +0x3d
218948 github.com/influxdata/influxdb/tsdb/engine/tsm1.(*FileStore).Close(0xc00dea3110, 0x133e3e8, 0xc0003b3440)
218949     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/file_store.go:604 +0xd0
218950 github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).Close(0xc0003b3440, 0x0, 0x0)
218951     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:768 +0xa0
218952 github.com/influxdata/influxdb/tsdb.(*Shard).close(0xc0074dc240, 0x133e3e8, 0xc0074dc3f8)
218953     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/shard.go:378 +0x47
218954 github.com/influxdata/influxdb/tsdb.(*Shard).Close(0xc0074dc240, 0x0, 0x0)
218955     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/shard.go:368 +0x79
218956 github.com/influxdata/influxdb/tsdb.(*Store).DeleteShard(0xc000185000, 0x556, 0x0, 0x0)
218957     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/tsdb/store.go:780 +0x317
218958 github.com/influxdata/influxdb/coordinator.(*StatementExecutor).executeDropShardStatement(0xc0051151f0, 0xc18eb83a70, 0x0, 0x48)
218959     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/coordinator/statement_executor.go:377 +0x45
218960 github.com/influxdata/influxdb/coordinator.(*StatementExecutor).ExecuteStatement(0xc0051151f0, 0x146e340, 0xc18eb83a70, 0xc1d92a7200, 0x1, 0x1)
218961     /Users/suhai/Code/go/src/github.com/influxdata/influxdb/coordinator/statement_executor.go:132 +0x1aa9
218962 github.com/influxdata/influxdb/query.(*Executor).executeQuery(0xc000572a20, 0xc1dbe9d740, 0xc031459219, 0x0, 0x0, 0x0, 0x146f980, 0x1ef6858, 0x2710, 0x       0, ...)
```

**Comment by GenuineJyn** [comment]
Yesterday i read the code `tsdb/engine/tsm1/reader.go` , `tsdb/engine/tsm1/file_store.go`, `tsdb/engine/tsm1/engine.go` To Find the pair`Ref()`and `Unref()`，when queries coming，newKeyCursor would be called , where the Ref() would be called, there are too many codes for query operation，the Unref() may not be called for some reason.

**Comment by GenuineJyn** [comment]
I got it.  My colleagues have implemented a cluster version with v1.6.1.
when using meta.Client to send query to several peers(RPCServer) , not all succed and forget closeing the successful return cursors.
